src/components/Microsoft.DotNet.UpgradeAssistant.MSBuild/PackageLoader.cs

```diff
@@ -141,31 +119,59 @@ public async Task<IEnumerable<NuGetReference>> GetNewerVersionsAsync(NuGetRefere
-            {
-                return null;
-            }
-
-            return new NuGetReference(packageName, highestVersion.ToFullString());
+            return FilterSearchResults(name, results, tfms, currentVersion, latestMinorAndBuildOnly);
         }
 
-        private List<PackageSource> GetPackageSources(string? projectDir)
+        public static IEnumerable<NuGetReference> FilterSearchResults(
+            string name,
+            IReadOnlyCollection<IPackageSearchMetadata> searchResults,
+            IEnumerable<TargetFrameworkMoniker> tfms,
+            NuGetVersion? currentVersion = null,
+            bool latestMinorAndBuildOnly = false)
         {
-            var packageSources = new List<PackageSource>();
-            if (projectDir != null)
+            if (searchResults is null || searchResults.Count == 0)
             {
-                var nugetSettings = Settings.LoadDefaultSettings(projectDir);
-                var sourceProvider = new PackageSourceProvider(nugetSettings);
-                packageSources.AddRange(sourceProvider.LoadPackageSources().Where(e => e.IsEnabled));
+                return Enumerable.Empty<NuGetReference>();
             }
 
-            if (packageSources.Count == 0)
+            var tfmSet = ImmutableHashSet.CreateRange(tfms.Select(t => NuGetFramework.Parse(t.Name)));
+
+            var results = searchResults
+                .Where(r => currentVersion is null ? true : r.Identity.Version > currentVersion);
+
+            if (latestMinorAndBuildOnly)
             {
-                packageSources.Add(new PackageSource(DefaultPackageSource));
+                results = results
+                    .GroupBy(r => r.Identity.Version.Major)
+                    .SelectMany(r =>
+                    {
+                        var max = r.Max(t => t.Identity.Version);
+
+                        return r.Where(t => t.Identity.Version == max);
+                    });
             }
 
-            _logger.LogDebug("Found package sources: {PackageSources}", packageSources);
-            return packageSources;
+            return results
+                .Where(r =>
+                {
+                    var unsupported = tfmSet;
+
+                    foreach (var dep in r.DependencySets)
+                    {
+                        foreach (var t in unsupported)
+                        {
+                            if (DefaultCompatibilityProvider.Instance.IsCompatible(t, dep.TargetFramework))
+                            {
+                                unsupported = unsupported.Remove(t);
+                            }
+                        }
+                    }
+
+                    return unsupported.IsEmpty;
+                })
+                .Select(r => r.Identity.Version)
+                .OrderBy(v => v)
+                .Select(v => new NuGetReference(name, v.ToNormalizedString()));
         }
 
         private async Task<T> CallWithRetryAsync<T>(Func<Task<T>> func)
```

Comment:
I think we should start thinking about bringing in polly. I don't think it holds up this PR

Reply:
I agree. I started breaking things apart and almost added polly, but I was able to add testing to a static method, so I didn't continue down that path.